src/linked_list.js

```diff
@@ -0,0 +1,136 @@
+'use strict'
+
+function Node( data ) {
+  this._data = data
+  this._next = null
+}
+
+export default class LinkedList {
+  constructor() {
+    this._tail = null
+    this._head = null
+    this._size = 0
+  }
+
+  size() {
+    return this._size
+  }
+
+  isEmpty() {
+    return this._size == 0
+  }
+
+  clear() {
+    this._tail = null
+    this._head = null
+    this._size = 0
+  }
+
+  getHeadNode() {
+    if ( this._head ) { return this._head }
+    else { return null }
+  }
+
+  getTailNode() {
+    if ( this._tail ) { return this._tail }
+    else { return null }
+  }
+
+  insert( data ) {
+    const newNode = new Node( data )
+    if ( this._size === 0 ) {
+      this._tail = newNode
+    } else {
+      newNode._next = this._head
+    }
+    this._head = newNode
+    ++this._size
+  }
+
+  insertFirst( data ) {
+    const newNode = new Node( data )
+    if ( this._size === 0 ) {
+      this._head = newNode
+    } else {
+      const oldTail = this._tail
+      oldTail._next = newNode
+    }
+    this._tail = newNode
+    ++this._size
+  }
+
+  remove() {
+    if ( this.isEmpty() ) { return null }
+    const oldHead = this._head
+    --this._size
+    let newHead = oldHead.next
+    this._head = newHead
+    return oldHead._data
+  }
+
+  removeFirst() {
+    if ( this.isEmpty() ) { return null }
+    const oldTail = this._tail
+    --this._size
+    let newFront = this._head
+    for ( const i = 1; i < this._size; ++i) {
+      newFront = newFront._next
+    }
+    newFront._next = null
+    this._tail = newFront
+    return oldTail._data
+  }
```

Comment:
I can see the benefits of keeping track of the tail. But there seems to be more code keeping track of it that it actually saves.